# Regression Model in Keras

## Part A

A. Build a baseline model (5 marks)

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [0]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
df.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

In [3]:
df.shape

(1030, 9)

In [0]:
columns = df.columns
X = df[columns[columns != 'Strength']] # all columns except Strength
y = df['Strength'] # Strength column

In [18]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from keras.models import Sequential
from keras.layers import Dense

MSE = []

for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.30, random_state = 1)
    n_cols = X_train.shape[1]
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, validation_data = (X_test, y_test), epochs = 50, verbose = 0)
    
    scores = model.evaluate(X_test, y_test, verbose = 0)
    print('Iteration {} evaluation : {}'.format(i + 1, scores))
    
    pred = model.predict(X)
    mse = mean_squared_error(y, pred)
    print('Iteration {} MSE : {}'.format(i + 1, scores))
    MSE.append(mse)

Iteration 1 evaluation : 140.56242923674847
Iteration 1 MSE : 140.56242923674847
Iteration 2 evaluation : 120.3386398364811
Iteration 2 MSE : 120.3386398364811
Iteration 3 evaluation : 974.9368955741809
Iteration 3 MSE : 974.9368955741809
Iteration 4 evaluation : 123.59934538313486
Iteration 4 MSE : 123.59934538313486
Iteration 5 evaluation : 96.94179028754867
Iteration 5 MSE : 96.94179028754867
Iteration 6 evaluation : 159.67372230110044
Iteration 6 MSE : 159.67372230110044
Iteration 7 evaluation : 380.9112148840451
Iteration 7 MSE : 380.9112148840451
Iteration 8 evaluation : 1526.2388872560175
Iteration 8 MSE : 1526.2388872560175
Iteration 9 evaluation : 296.8929666562374
Iteration 9 MSE : 296.8929666562374
Iteration 10 evaluation : 1529.989217912495
Iteration 10 MSE : 1529.989217912495
Iteration 11 evaluation : 121.89282154959768
Iteration 11 MSE : 121.89282154959768
Iteration 12 evaluation : 223.39409368323663
Iteration 12 MSE : 223.39409368323663
Iteration 13 evaluation : 200.2544

In [25]:
MSE = np.array(MSE)

mean_stepA = MSE.mean()
std_step_A = MSE.std()
print(mean_stepA)
print(std_step_A)

336.1609105416868
397.909409251028


## Part B

B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [8]:
X_norm = (X - X.mean()) / X.std()
X_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [20]:
MSE = []

for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.30, random_state = 1)
    n_cols = X_train.shape[1]
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, validation_data = (X_test, y_test), epochs = 50, verbose = 0)
    
    scores = model.evaluate(X_test, y_test, verbose = 0)
    print('Iteration {} evaluation : {}'.format(i + 1, scores))
    
    pred = model.predict(X_norm)
    mse = mean_squared_error(y, pred)
    print('Iteration {} MSE : {}'.format(i + 1, scores))
    MSE.append(mse)

Iteration 1 evaluation : 481.04590475829286
Iteration 1 MSE : 481.04590475829286
Iteration 2 evaluation : 285.61921291906856
Iteration 2 MSE : 285.61921291906856
Iteration 3 evaluation : 290.1271908460697
Iteration 3 MSE : 290.1271908460697
Iteration 4 evaluation : 446.0973953197689
Iteration 4 MSE : 446.0973953197689
Iteration 5 evaluation : 282.2130398549694
Iteration 5 MSE : 282.2130398549694
Iteration 6 evaluation : 684.2799048562651
Iteration 6 MSE : 684.2799048562651
Iteration 7 evaluation : 295.82480757136176
Iteration 7 MSE : 295.82480757136176
Iteration 8 evaluation : 384.39704929586367
Iteration 8 MSE : 384.39704929586367
Iteration 9 evaluation : 353.3376802611119
Iteration 9 MSE : 353.3376802611119
Iteration 10 evaluation : 284.5420151003742
Iteration 10 MSE : 284.5420151003742
Iteration 11 evaluation : 346.15242177228714
Iteration 11 MSE : 346.15242177228714
Iteration 12 evaluation : 353.8192261137237
Iteration 12 MSE : 353.8192261137237
Iteration 13 evaluation : 330.994539

In [23]:
MSE = np.array(MSE)

mean_stepB = MSE.mean()
std_stepB = MSE.std()
print(mean_stepB)
print(std_stepB)

343.36536222715586
82.66475866614807


### Comparison of step A and B :  
For step A, we had :  
- Mean = 336.1609105416868
- Standard deviation = 397.909409251028

And for step B :
- Mean = 343.36536222715586
- Standard deviation = 82.66475866614807

**The MSE is almost the same as before**

## Part C

C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [26]:
MSE = []

for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.30, random_state = 1)
    n_cols = X_train.shape[1]
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, validation_data = (X_test, y_test), epochs = 100, verbose = 0)
    
    scores = model.evaluate(X_test, y_test, verbose = 0)
    print('Iteration {} evaluation : {}'.format(i + 1, scores))
    
    pred = model.predict(X_norm)
    mse = mean_squared_error(y, pred)
    print('Iteration {} MSE : {}'.format(i + 1, scores))
    MSE.append(mse)

Iteration 1 evaluation : 177.3231744858825
Iteration 1 MSE : 177.3231744858825
Iteration 2 evaluation : 186.71338379498823
Iteration 2 MSE : 186.71338379498823
Iteration 3 evaluation : 171.86973319933253
Iteration 3 MSE : 171.86973319933253
Iteration 4 evaluation : 195.73187581775258
Iteration 4 MSE : 195.73187581775258
Iteration 5 evaluation : 203.2391630993692
Iteration 5 MSE : 203.2391630993692
Iteration 6 evaluation : 184.40308081370728
Iteration 6 MSE : 184.40308081370728
Iteration 7 evaluation : 185.4972838861657
Iteration 7 MSE : 185.4972838861657
Iteration 8 evaluation : 180.7276218253818
Iteration 8 MSE : 180.7276218253818
Iteration 9 evaluation : 222.27923623489332
Iteration 9 MSE : 222.27923623489332
Iteration 10 evaluation : 199.62125809061487
Iteration 10 MSE : 199.62125809061487
Iteration 11 evaluation : 212.0272833074181
Iteration 11 MSE : 212.0272833074181
Iteration 12 evaluation : 178.69214736296522
Iteration 12 MSE : 178.69214736296522
Iteration 13 evaluation : 183.04

In [27]:
MSE = np.array(MSE)

mean_stepC = MSE.mean()
std_stepC = MSE.std()
print(mean_stepC)
print(std_stepC)

164.06538841617134
16.388811252309374


### Comparison of step B and C :  
For step B, we had :  
- Mean = 343.36536222715586
- Standard deviation = 82.66475866614807

And for step C :
- Mean = 164.06538841617134
- Standard deviation = 16.388811252309374

**The MSE is almost 1/2 of before, meaning that the model is better**

## Part D

D. Increase the number of hidden layers (5 marks)

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [28]:
MSE = []

for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size = 0.30, random_state = 1)
    n_cols = X_train.shape[1]
    model = Sequential()
    model.add(Dense(10, activation = 'relu', input_shape = (n_cols,)))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(1))

    # compile model
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    model.fit(X_train, y_train, validation_data = (X_test, y_test), epochs = 50, verbose = 0)
    
    scores = model.evaluate(X_test, y_test, verbose = 0)
    print('Iteration {} evaluation : {}'.format(i + 1, scores))
    
    pred = model.predict(X_norm)
    mse = mean_squared_error(y, pred)
    print('Iteration {} MSE : {}'.format(i + 1, scores))
    MSE.append(mse)

Iteration 1 evaluation : 151.27031186792073
Iteration 1 MSE : 151.27031186792073
Iteration 2 evaluation : 145.29680586632787
Iteration 2 MSE : 145.29680586632787
Iteration 3 evaluation : 140.9290121627857
Iteration 3 MSE : 140.9290121627857
Iteration 4 evaluation : 154.9971687292204
Iteration 4 MSE : 154.9971687292204
Iteration 5 evaluation : 140.83912135250745
Iteration 5 MSE : 140.83912135250745
Iteration 6 evaluation : 139.35664856549604
Iteration 6 MSE : 139.35664856549604
Iteration 7 evaluation : 145.2827419540257
Iteration 7 MSE : 145.2827419540257
Iteration 8 evaluation : 144.58277809118377
Iteration 8 MSE : 144.58277809118377
Iteration 9 evaluation : 128.82948436551882
Iteration 9 MSE : 128.82948436551882
Iteration 10 evaluation : 156.77977597134785
Iteration 10 MSE : 156.77977597134785
Iteration 11 evaluation : 155.3609023603421
Iteration 11 MSE : 155.3609023603421
Iteration 12 evaluation : 153.18323771158853
Iteration 12 MSE : 153.18323771158853
Iteration 13 evaluation : 144.

In [29]:
MSE = np.array(MSE)

mean_stepD = MSE.mean()
std_stepD = MSE.std()
print(mean_stepD)
print(std_stepD)

124.26739708633573
13.61897444955887


### Comparison of step B and D :  
For step B, we had :  
- Mean = 343.36536222715586
- Standard deviation = 82.66475866614807

And for step D :
- Mean = 124.26739708633573
- Standard deviation = 13.61897444955887

**The MSE is lower than step B**